pkg/pillar/cmd/client/client.go

```diff
@@ -516,42 +575,98 @@ func selfRegister(zedcloudCtx zedcloud.ZedCloudContext, tlsConfig *tls.Config, d
+	var resp *http.Response
+	var b, contents []byte
+	var done bool
+
+	if !force {
+		_, err := os.Stat(types.ServerCertName)
+		if err == nil {
+			return true, false
+		}
+	}
+
+	// certs API is always V2, and without UUID, use http for now
+	requrl := zedcloud.URLPathString(serverNameAndPort, true, true, nilUUID, "certs")
+	// currently there is no data included for the request, same as myGet()
+	done, resp, _, contents = myPost(zedcloudCtx, tlsConfig, requrl, retryCount, int64(len(b)), bytes.NewBuffer(b))
+	if resp != nil {
+		log.Infof("client fetchCertChain done %v, resp-code %d, content len %d", done, resp.StatusCode, len(contents))
+		if resp.StatusCode == http.StatusNotFound || resp.StatusCode == http.StatusUnauthorized ||
+			resp.StatusCode == http.StatusNotImplemented || resp.StatusCode == http.StatusBadRequest {
+			// cloud server does not support V2 API
+			log.Infof("client fetchCertChain: server %s does not support V2 API\n", serverName)
+			return false, true
+		}
+	} else {
+		log.Infof("client fetchCertChain done %v, resp null, content len %d", done, len(contents))
+	}
+	if !done {
+		return false, false
+	}
+
+	zedcloudCtx.TlsConfig = tlsConfig
+	// verify the certificate chain and write the siging cert to file
+	err := zedcloud.VerifyCloudCertChain(zedcloudCtx, serverName, contents)
+	if err != nil {
+		log.Infof("client fetchCertChain: verify err %v\n", err)
+		return false, false
+	}
+
+	log.Infof("client fetchCertChain: ok\n")
+	return true, false
 }
 
 func doGetUUID(zedcloudCtx zedcloud.ZedCloudContext, tlsConfig *tls.Config,
-	retryCount int) (bool, uuid.UUID, string, string, string) {
+	retryCount int) (bool, bool, uuid.UUID, string, string, string) {
 
 	var resp *http.Response
 	var contents []byte
+	var rtf types.HTTPExtraStatus
 
-	requrl := serverNameAndPort + "/api/v1/edgedevice/config"
+	// get UUID does not have UUID string in V2 API
+	requrl := zedcloud.URLPathString(serverNameAndPort, zedcloudCtx.V2API, false, nilUUID, "config")
 	b, err := generateConfigRequest()
 	if err != nil {
 		log.Errorln(err)
-		return false, nilUUID, "", "", ""
+		return false, false, nilUUID, "", "", ""
 	}
 	var done bool
-	done, resp, contents = myPost(zedcloudCtx, tlsConfig, requrl, retryCount,
+	done, resp, rtf, contents = myPost(zedcloudCtx, tlsConfig, requrl, retryCount,
 		int64(len(b)), bytes.NewBuffer(b))
 	if resp != nil && resp.StatusCode == http.StatusNotModified {
 		// Acceptable response for a ConfigRequest POST
 		done = true
 	}
 	if !done {
-		return false, nilUUID, "", "", ""
+		// This may be due to the cloud cert file is stale, since the hash does not match.
+		// acquire new cert chain.
+		if rtf == types.ExtStatusCertMiss {
+			var gotServerCerts bool
+			gotServerCerts, _ = fetchCertChain(zedcloudCtx, tlsConfig, retryCount, true)
+			log.Infof("doGetUUID: got server certs %v\n", gotServerCerts)
+		}
```

Comment:
Better to set a triggerFetchCerts boolean (in clientContext, or global) and have the main event loop look at that to do the fetching. That way we will have a better handle on how long time (how many http/tcp timeouts) we might spend from the top of the event loop. And the code in the event loop to fetch the initial certs and the triggered ones will be (more) common.

Reply:
ok, i can convert to a timer to trigger here.